alphazero/mcts/mcts.py

```diff
@@ -0,0 +1,282 @@
+from __future__ import annotations
+from typing import List, TypeVar
+import numpy as np
+import torch
+from alphazero import TicTacToe
+from alphazero.utils import sample_tensor_indices
+
+GameClassType = TypeVar("T", bound="Game")
+
+
+class BoardConverter:
+    @staticmethod
+    def board_to_tensor(board):
+        raise NotImplementedError
+
+    @staticmethod
+    def action_to_tensor(action):
+        # Not sure if needed
+        raise NotImplementedError
+
+    @staticmethod
+    def tensor_to_action(action_tensor):
+        raise NotImplementedError
+
+
+class TicTacToeConverter(BoardConverter):
+    @staticmethod
+    def board_to_tensor(board):
+        player_1 = board == 0
+        player_2 = board == 1
+        player = 1 if player_1.sum() > player_2.sum() else 0
+        player_layer = torch.full(board.size(), player)
+        return torch.stack((player_1, player_2, player))
+
+    ## TODO: Flesh this out and move
+
+
+class MCTSNode:
+    """
+    Node Class for MCTS tree search
+    """
+
+    def __init__(
+        self,
+        state: torch.Tensor,
+        prior: float = 0,
+        action: torch.Tensor = torch.Tensor([]),
+    ):
+
+        self._action = action
+        self._state = state
+        self._children: List[MCTSNode] = []
+        self._value = None
+        self._policy = None
+        self.n_visit = 0
+        self.total_value = 0
+        self.is_terminal = False
+        self.prior = prior
+        self.parent = None
+
+    def add_child(self, child: MCTSNode):
+        """
+        Adds a child to node instance
+
+        Args:
+            child (MCTSNode): Node to be added as child
+        """
+        self._children.append(child)
+        child.parent = self
+
+    def rollout(self, game: GameClassType) -> int:
+        """
+        Performs a rollout from self, selecting random actions from a uniform prior until game ends
+        Args:
+            game (GameClassType): Game that is currently played
+
+        Returns:
+            result (int): Final result of rollout
+        """
+        state = self._state
+        while not game.is_game_over(state):
+            legal_actions = game.get_legal_moves(state)
+            action = sample_tensor_indices(legal_actions, 1)[0]
+            player = 1 if (state == 0).sum() > (state == 1).sum() else 0
+            state = game.board_after_move(state, player, action)
+        result = game.result(state)
+        return result
+
+    def calc_policy_value(
+        self, rollout: bool, game: GameClassType, network: torch.nn.Module = None
+    ) -> (torch.Tensor, float):
+        """
+        Calculates the policy and value for the node. If rollout performs a rollout, else uses network to get policy and value
```

Comment:
Can you separate this into two lines?

(this is one of those things that black doesn't do for you... I think we should keep to a 80 char limit (or so) per line. It helps me read code with my tiny 768p screen :P )

Reply:
Haha yeah, I actually got a linting warning but ignored it, my bad